In [1]:
import pandas as pd
import numpy as np

In [2]:
df=pd.read_csv('/content/PR4_STUDENTS_PERFORMANCE.CSV')

In [3]:
df

,STUDENT_ID,FIRST_NAME,LAST_NAME,FULL_NAME,MATH,SCIENCE,ENGLISH,HISTORY,TOTAL_SCORE,AVERAGE_SCORE,GRADE,STATUS
0,S001,Aarav,Sharma,Aarav Sharma,73,66,61,62,262,65.50,C,PASS
1,S002,Vivaan,Patel,Vivaan Patel,71,83,95,79,328,82.00,A,PASS
2,S003,Aditya,Verma,Aditya Verma,69,78,69,66,282,70.50,B,PASS
3,S004,Arjun,Singh,Arjun Singh,87,85,80,74,326,81.50,A,PASS
4,S005,Reyansh,Khan,Reyansh Khan,91,72,88,70,321,80.25,A,PASS
...,...,...,...,...,...,...,...,...,...,...,...,...
195,S196,Sneha,Pandya,Sneha Pandya,79,76,88,82,325,81.25,A,PASS
196,S197,Tanya,Sharma,Tanya Sharma,73,77,79,69,298,74.50,B,PASS
197,S198,Meera,Patel,Meera Patel,53,60,56,61,230,57.50,D,PASS
198,S199,Aditi,Verma,Aditi Verma,78,60,63,68,269,67.25,C,PASS


**part A  VECTOR AND MATRIX FUNDAMENTALS**

# 1.REPRESENT EACH STUDENTS SUBJECTS AS VECTOR.

In [4]:
subjects = ["MATH", "SCIENCE", "ENGLISH", "HISTORY"]

A = df[subjects].to_numpy()

print("S001 Vector:", A[0])
print("S002 Vector:", A[1])

S001 Vector: [73 66 61 62]
S002 Vector: [71 83 95 79]


In [5]:
for i in range(len(df)):
    print(df.loc[i, "STUDENT_ID"], "=", A[i])

S001 = [73 66 61 62]
S002 = [71 83 95 79]
S003 = [69 78 69 66]
S004 = [87 85 80 74]
S005 = [91 72 88 70]
S006 = [76 58 72 64]
S007 = [53 59 71 79]
S008 = [74 76 71 75]
S009 = [56 64 76 71]
S010 = [92 76 83 90]
S011 = [57 75 64 65]
S012 = [90 94 92 89]
S013 = [74 69 64 79]
S014 = [82 78 75 81]
S015 = [73 85 73 62]
S016 = [80 80 78 70]
S017 = [80 95 89 81]
S018 = [61 56 83 68]
S019 = [75 83 91 80]
S020 = [76 72 85 79]
S021 = [66 83 96 78]
S022 = [77 60 80 63]
S023 = [80 64 77 78]
S024 = [68 73 90 62]
S025 = [69 72 80 83]
S026 = [79 74 78 68]
S027 = [59 60 73 72]
S028 = [74 88 88 74]
S029 = [69 75 74 72]
S030 = [76 82 74 74]
S031 = [79 65 67 69]
S032 = [80 71 86 74]
S033 = [74 74 81 79]
S034 = [78 58 72 66]
S035 = [66 59 71 69]
S036 = [89 80 83 85]
S037 = [86 84 75 71]
S038 = [51 46 60 59]
S039 = [76 76 77 77]
S040 = [47 46 56 50]
S041 = [76 94 85 79]
S042 = [73 65 75 75]
S043 = [84 76 72 73]
S044 = [94 88 82 85]
S045 = [61 61 67 72]
S046 = [58 52 61 65]
S047 = [67 58 68 65]
S048 = [79 68

**2.  compute norm-1 and norm-2 of vectors**

2.1
  NORM-1
   v1=[73,66,61,62]
    v1=[73+66+61+62]
    v1=[262]


NORM-2
   V2={ $$ V2= \sqrt{73^2+66^2+61^2+62^2} $$ $$ =\sqrt{5329+4356+3721+3844} $$ $$ =\sqrt{17250} $$ $$ \boxed{V2=131.34} $$

In [6]:
v = A[0]

norm_1 = np.linalg.norm(v, ord=1)
norm_2 = np.linalg.norm(v, ord=2)

print("S001 Vector:", v)
print("Norm-1:", norm_1)
print("Norm-2:", round(norm_2, 2))

S001 Vector: [73 66 61 62]
Norm-1: 262.0
Norm-2: 131.34


2.2
 Dot Product and Angle between two students' score vectors

 dot product of 2 students

 Vectors
 V=[73,66,61,62]
W=[71,83,95,79]

V⋅W=v1​w1​+v2​w2​+v3​w3​+v4​w4​

=(73)(71)+(66)(83)+(61)(95)+(62)(79)
$$ =5183+5478+5795+4898 $$
V⋅W=21354
	​



In [7]:
v = A[0]
w = A[1]

dot_product = np.dot(v, w)

print("Dot Product:", dot_product)

Dot Product: 21354


## angle between vectors


In [8]:
angle_rad = np.arccos(
    np.clip(
        np.dot(v, w) /
        (np.linalg.norm(v) * np.linalg.norm(w)),
        -1, 1
    )
)

angle_deg = np.degrees(angle_rad)

print("Angle:", round(angle_deg, 2), "degrees")

Angle: 9.63 degrees


# 2.3 Cross Product — 3D selected subjects

## cross product sirf 3 dimensional vectors ke liya hota hai

MATH
SCIENCE
ENGLISH

S1  V=[73,66,61]
S2  W=[71,83,95]

V×W=
​i7371

​j6683

​k6195​

RESULT

[1207,-2604,1373]
​

In [9]:
v3 = A[0, :3]
w3 = A[1, :3]

cross_product = np.cross(v3, w3)

print("S001 3D Vector:", v3)
print("S002 3D Vector:", w3)
print("Cross Product:", cross_product)

S001 3D Vector: [73 66 61]
S002 3D Vector: [71 83 95]
Cross Product: [ 1207 -2604  1373]


**3. Find the projection of one vector INTO ANOTHER**

V=[73,66,61,62]

W=[71,83,95,79]

In [10]:
projection = (
    np.dot(v, w) /
    np.dot(w, w)
) * w

print("Projection of S001 onto S002:")
print(np.round(projection, 2))

Projection of S001 onto S002:
[55.75 65.17 74.59 62.03]


**PART B MATRIX OPERATION**
## 4.1 MATRIX ADDITION
A1= STUDENTS 1-100
A2=STUDENTS 101-200

BOTH SIZE IS   

100*4

In [11]:
A1 = A[:100]
A2 = A[100:]

addition = A1 + A2

print("A1 Shape:", A1.shape)
print("A2 Shape:", A2.shape)
print("Addition Shape:", addition.shape)

print("First 5 rows of A1 + A2:")
print(addition[:5])

A1 Shape: (100, 4)
A2 Shape: (100, 4)
Addition Shape: (100, 4)
First 5 rows of A1 + A2:
[[155 143 148 150]
 [150 163 170 161]
 [146 168 146 135]
 [168 166 162 154]
 [144 124 150 126]]


FIRST ROWS
[73,66,61,62]+[82,77,87,88]
=

# RESULT = [155,143,148,150]
	​

SO, MATRIX ADDITION IS PERFORMED ELEMENT BY ELEMENT.

MATRIX MULTIPLICATION

Original matrix:

$$ A=200\times4 $$

Agar:

$$ A\times A $$

karenge, dimensions hongi:

$$ (200\times4)(200\times4) $$

Ye possible nahi hai because inner dimensions:

$$ 4\neq200 $$

Isliye hum transpose use karenge.

$$ A^T=4\times200 $$

Then:

$$ A^T A $$ $$ =(4\times200)(200\times4) $$
=
# RESULT 4×4
	​


In [12]:
A_T = A.T

multiplication = A_T @ A

print("Original Matrix Shape:", A.shape)
print("Transpose Shape:", A_T.shape)
print("Multiplication Result Shape:", multiplication.shape)

print("Aᵀ × A:")
print(multiplication)

Original Matrix Shape: (200, 4)
Transpose Shape: (4, 200)
Multiplication Result Shape: (4, 4)
Aᵀ × A:
[[1135101 1095754 1150402 1120794]
 [1095754 1077685 1121251 1091080]
 [1150402 1121251 1184409 1146862]
 [1120794 1091080 1146862 1124534]]


## 4.2 TRANSPODE AND INVERSE ( IF POSSIBLE)

In [14]:
subjects = ["MATH", "SCIENCE", "ENGLISH", "HISTORY"]

# Students × Subjects Matrix
A = df[subjects].to_numpy()

# Transpose
A_T = A.T
print("Original:", A.shape)
print("Transpose:", A_T.shape)

# Inverse possible? Original 200×4 hai, isliye NO
print("Inverse of original matrix: Not possible (200×4 is not square)")

# 4×4 matrix se inverse
A_square = A[:4, :4]
det = np.linalg.det(A_square)

print("4×4 Determinant:", round(det, 2))

if not np.isclose(det, 0):
    A_inv = np.linalg.inv(A_square)
    print("Inverse:")
    print(np.round(A_inv, 4))

Original: (200, 4)
Transpose: (4, 200)
Inverse of original matrix: Not possible (200×4 is not square)
4×4 Determinant: 102708.0
Inverse:
[[-0.002  -0.0183 -0.0697  0.0834]
 [-0.0965 -0.0375  0.0943  0.0368]
 [-0.0825  0.0321 -0.0828  0.1087]
 [ 0.2023  0.0299  0.0632 -0.2443]]


4.3 Dataset ka 200 × 4 matrix hai, isliye direct determinant possible nahi hai.

4 students × 4 subjects ka square matrix lete hain:

A = [73 66 61 62,
     71 83 95 79,
     69 78 69 66,
     87 85 80 74]

In [16]:
A = df[["MATH","SCIENCE","ENGLISH","HISTORY"]].iloc[:4].to_numpy()

det = np.linalg.det(A)

print("4×4 Matrix:")
print(A)
print("Determinant =", round(det, 2))
print("Since determinant ≠ 0, inverse of this 4×4 matrix is possible.")

4×4 Matrix:
[[73 66 61 62]
 [71 83 95 79]
 [69 78 69 66]
 [87 85 80 74]]
Determinant = 102708.0
Since determinant ≠ 0, inverse of this 4×4 matrix is possible.


** part c**

 5 . dataset has 4 subject dimensions:

MATH, SCIENCE, ENGLISH, HISTORY

So each student is represented as a 4D vector:

$$ Student = [Math,\ Science,\ English,\ History] $$

Example:

$$ S001=[73,66,61,62] $$
1. Line

A line is 1-dimensional.

In your dataset, if we vary only one direction, for example overall performance scaling:

$$ T(x)=x[73,66,61,62] $$

This represents a line through the origin in the 4D student-performance space.

2. Plane

A plane is 2-dimensional.

Take two students:

$$ S001=[73,66,61,62] $$ $$ S002=[71,83,95,79] $$

Their linear combinations

$$ aS001+bS002 $$

form a 2D plane inside the 4D dataset space.

3. Hyperplane

A hyperplane in 4D is 3-dimensional.

Using three independent student vectors:

$$ aS001+bS002+cS003 $$

creates a 3D hyperplane/subspace inside the 4D subject space.


conclusion: dataset has 4 dimensions, so students exist in a 4D feature space. Line = 1D, Plane = 2D, Hyperplane = 3D.

# 6  Dimensionality: 2D → 3D → Higher Dimensions

Hyperplane concept
2D: A line separates/represents points.
3D: A plane separates/represents points.
4D: A 3D hyperplane is used.
n dimensions: Hyperplane has n − 1 dimensio

In [17]:
subjects = ["MATH", "SCIENCE", "ENGLISH", "HISTORY"]

X_2D = df[["MATH","SCIENCE"]]
X_3D = df[["MATH","SCIENCE","ENGLISH"]]
X_4D = df[subjects]

print("2D:", X_2D.shape)
print("3D:", X_3D.shape)
print("4D:", X_4D.shape)
print("4D Hyperplane dimension =", X_4D.shape[1] - 1)

2D: (200, 2)
3D: (200, 3)
4D: (200, 4)
4D Hyperplane dimension = 3


**PART D**

# 7. Eigenvalues & Decomposition
# Compute the eigenvalues and eigenvectors of the covariance matrix.
1. Eigenvalue kya hota hai?

Eigenvalue batata hai ki covariance matrix mein kis direction mein data ka variance/spread kitna hai.

Eigenvector batata hai ki woh direction kya hai.

2. Covariance Matrix

Using all 200 students:

$$ C= \begin{bmatrix} 122.14&71.22&58.97&65.60\\ 71.22&123.36&62.97&62.68\\ 58.97&62.97&86.28&56.92\\ 65.60&62.68&56.92&99.71 \end{bmatrix} $$
3. Eigenvalues
$$ \boxed{ 299.149,\quad52.325,\quad45.722,\quad34.292 } $$
4. Eigenvectors

Corresponding eigenvectors:

[-0.539, -0.543, -0.436, -0.473]

[-0.644,  0.729,  0.110, -0.205]

[ 0.542,  0.335, -0.338, -0.692]

[-0.023,  0.247, -0.826,  0.506]

In [18]:
X = df[["MATH","SCIENCE","ENGLISH","HISTORY"]].dropna().to_numpy()

C = np.cov(X, rowvar=False)

eigenvalues, eigenvectors = np.linalg.eigh(C)

idx = np.argsort(eigenvalues)[::-1]
eigenvalues = eigenvalues[idx]
eigenvectors = eigenvectors[:, idx]

print("Covariance Matrix:\n", np.round(C,2))
print("\nEigenvalues:", np.round(eigenvalues,3))
print("\nEigenvectors:\n", np.round(eigenvectors,3))

Covariance Matrix:
 [[122.14  71.22  58.97  65.6 ]
 [ 71.22 123.36  62.97  62.68]
 [ 58.97  62.97  86.28  56.92]
 [ 65.6   62.68  56.92  99.71]]

Eigenvalues: [299.149  52.325  45.722  34.292]

Eigenvectors:
 [[-0.539 -0.644  0.542 -0.023]
 [-0.543  0.729  0.335  0.247]
 [-0.436  0.11  -0.338 -0.826]
 [-0.473 -0.205 -0.692  0.506]]


**conclusion: Sabse bada eigenvalue 299.149 hai, matlab dataset ki sabse large variation first principal direction mein hai**.

In [20]:
print("🎓 LINEAR ALGEBRA ANALYSIS COMPLETED!")
print("✨ Thank You!")

🎓 LINEAR ALGEBRA ANALYSIS COMPLETED!
✨ Thank You!
